### PR TITLE
DirectionRepositoryTest

### PR DESCRIPTION
**Pull request by darkpchela**
Main file is DirectionRepositoryTest.cs. Is it made correctly?

### CODE REVIEW

File: DIMS-Core/DIMS-Core.DataAccessLayer/Context/DIMSCoreDatabaseContext.cs [hunk 1]
```diff
@@ -5,13 +5,13 @@
 
 namespace DIMS_Core.DataAccessLayer.Context
 {
-    public partial class DIMSCoreDatabaseContext : DbContext
+    public partial class DIMSCoreDataBaseContext : DbContext
```
Comment (PushMeNow) [inline DIMS-Core/DIMS-Core.DataAccessLayer/Context/DIMSCoreDatabaseContext.cs:8]: Why did you change 'Database' with 'Data**B**ase'? Please don't pollute your PR useless changes.

File: DIMS-Core/DIMS-Core.Tests/DAL/DirectionRepositoryTest.cs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+        public DirectionRepositoryTest()
+        {
+            _DbSetList = new List<Direction>()
+            {
+                new Direction { DirectionId = 1, Name = ".net", Description = "none" },
+                new Direction { DirectionId = 2, Name = "java", Description = "none" },
+                new Direction { DirectionId = 3, Name = "front-end", Description = "none" },
+                new Direction { DirectionId = 4, Name = "salesforce", Description = "none" }
+            };
```
Comment (PushMeNow) [inline DIMS-Core/DIMS-Core.Tests/DAL/DirectionRepositoryTest.cs:28]: Move these data in method. They will be able to you in the future.

File: DIMS-Core/DIMS-Core.Tests/DAL/DirectionRepositoryTest.cs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+            _DbSetDirectionMock = MockHelper.CreateDbSetMock<Direction>(_DbSetList);
+            _DbMock = new Mock<DIMSCoreDataBaseContext>();
+            _DbMock.Setup(db => db.Set<Direction>()).Returns(_DbSetDirectionMock.Object);
+            _DbMock.Setup(db => db.Direction).Returns(_DbSetDirectionMock.Object);
+        }
+
+        [Test]
+        public async System.Threading.Tasks.Task CreateNewEntity()
```
Comment (PushMeNow) [inline DIMS-Core/DIMS-Core.Tests/DAL/DirectionRepositoryTest.cs:37]: Don't use full namespaces for calling of types. Use 'using'.

File: DIMS-Core/DIMS-Core.Tests/DAL/DirectionRepositoryTest.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        {
+            Direction direction = new Direction { Name = "C++", Description = "none" };
+            var repo = new DirectionRepository(_DbMock.Object);
+            await repo.CreateAsync(direction);
+            _DbSetDirectionMock.Verify(db => db.AddAsync(It.IsAny<Direction>(), It.IsAny<CancellationToken>()), Times.Once);
+        }
+
+        [Test]
+        [TestCase(1)]
+        public async System.Threading.Tasks.Task DeleteExistingElement(int id)
```
Comment (PushMeNow) [inline DIMS-Core/DIMS-Core.Tests/DAL/DirectionRepositoryTest.cs:47]: Don't use full namespaces for calling of types. Use 'using'.

File: DIMS-Core/DIMS-Core.Tests/DAL/DirectionRepositoryTest.cs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        public void GetAllFromRepository()
+        {
+            var repo = new DirectionRepository(_DbMock.Object);
+            var res = repo.GetAll();
+            Assert.IsTrue(res.Count() == 4);
+        }
+
+        [Test]
+        [TestCase(1)]
+        public async System.Threading.Tasks.Task GetByIdExisting(int id)
```
Comment (PushMeNow) [inline DIMS-Core/DIMS-Core.Tests/DAL/DirectionRepositoryTest.cs:65]: Don't use full namespaces for calling of types. Use 'using'.

File: DIMS-Core/DIMS-Core.Tests/DAL/DirectionRepositoryTest.cs [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+            _DbSetDirectionMock.Setup(m => m.FindAsync(It.IsAny<int>())).ReturnsAsync(() => { return _DbSetList.SingleOrDefault(d => d.DirectionId == id); });
+            var repo = new DirectionRepository(_DbMock.Object);
+            var res = await repo.GetByIdAsync(id);
+            Assert.AreEqual(_DbSetList.ElementAt(id - 1), res);
+        }
+
+        [Test]
+        [TestCase(-300)]
+        [TestCase(100)]
+        public async System.Threading.Tasks.Task GetByIdNotExisting(int id)
```
Comment (PushMeNow) [inline DIMS-Core/DIMS-Core.Tests/DAL/DirectionRepositoryTest.cs:76]: Don't use full namespaces for calling of types. Use 'using'.

File: DIMS-Core/DIMS-Core.Tests/DAL/TaskRepositoryTest.cs [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        private readonly Mock<DbSet<Task>> _DbSetTaskMock;
+
+        public TaskRepositoryTest()
+        {
+            _DbSetList = new List<Task>()
+            {
+                new Task { TaskId=1, Name="Create Unit tests", Description="none", StartDate=DateTime.Parse("20.07.2020"), DeadlineDate=DateTime.Parse("22.07.2020")},
+                new Task { TaskId=2, Name="Implement DAL", Description="none", StartDate=DateTime.Parse("19.07.2020"), DeadlineDate=DateTime.Parse("27.07.2020")},
+                new Task { TaskId=3, Name="Implement BLL", Description="none", StartDate=DateTime.Parse("23.07.2020"), DeadlineDate=DateTime.Parse("30.07.2020")},
+            };
```
Comment (PushMeNow) [inline DIMS-Core/DIMS-Core.Tests/DAL/TaskRepositoryTest.cs:28]: Move data in method. And don't use inline class initialization.

File: DIMS-Core/DIMS-Core.Tests/DAL/TaskRepositoryTest.cs [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        {
+            var repository = new TaskRepository(_DbMock.Object);
+            var res = repository.GetAll();
+            Assert.IsTrue(res.Count() == 3);
+        }
+
+        [Test]
+        public async System.Threading.Tasks.Task CreateNewEntity()
+        {
+            Task Task = new Task { Name = "Write report", Description = "none", StartDate = DateTime.Parse("23.07.2020"), DeadlineDate = DateTime.Parse("24.07.2020") };
```
Comment (PushMeNow) [inline DIMS-Core/DIMS-Core.Tests/DAL/TaskRepositoryTest.cs:47]: Don't use inline class initialization.

File: DIMS-Core/DIMS-Core.Tests/DAL/TaskRepositoryTest.cs [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        [Test]
+        public void GetAllFromRepository()
+        {
+            var repository = new TaskRepository(_DbMock.Object);
+            var res = repository.GetAll();
+            Assert.IsTrue(res.Count() == 3);
+        }
+
+        [Test]
+        public async System.Threading.Tasks.Task CreateNewEntity()
```
Comment (PushMeNow) [inline DIMS-Core/DIMS-Core.Tests/DAL/TaskRepositoryTest.cs:45]: namespace

File: DIMS-Core/DIMS-Core.Tests/DAL/TaskRepositoryTest.cs [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        {
+            Task Task = new Task { Name = "Write report", Description = "none", StartDate = DateTime.Parse("23.07.2020"), DeadlineDate = DateTime.Parse("24.07.2020") };
+            var repository = new TaskRepository(_DbMock.Object);
+            await repository.CreateAsync(Task);
+            _DbSetTaskMock.Verify(db => db.AddAsync(It.IsAny<Task>(), It.IsAny<CancellationToken>()), Times.Once);
+        }
+
+        [Test]
+        [TestCase(1)]
+        public async System.Threading.Tasks.Task DeleteExistingElement(int id)
```
Comment (PushMeNow) [inline DIMS-Core/DIMS-Core.Tests/DAL/TaskRepositoryTest.cs:55]: namespace

File: DIMS-Core/DIMS-Core.Tests/DAL/TaskRepositoryTest.cs [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+        {
+            _DbSetTaskMock.Setup(m => m.FindAsync(It.IsAny<int>())).ReturnsAsync(() => { return _DbSetList.SingleOrDefault(d => d.TaskId == id); });
+            var repository = new TaskRepository(_DbMock.Object);
+            await repository.DeleteAsync(id);
+            _DbSetTaskMock.Verify(m => m.Remove(It.IsAny<Task>()), Times.Once);
+        }
+
+        [Test]
+        [TestCase(-300)]
+        public async System.Threading.Tasks.Task GetByIdNotExisting(int id)
```
Comment (PushMeNow) [inline DIMS-Core/DIMS-Core.Tests/DAL/TaskRepositoryTest.cs:65]: namespace

File: DIMS-Core/DIMS-Core.Tests/DAL/TaskRepositoryTest.cs [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+        {
+            _DbSetTaskMock.Setup(m => m.FindAsync(It.IsAny<int>())).ReturnsAsync(() => { return _DbSetList.SingleOrDefault(d => d.TaskId == id); });
+            var repository = new TaskRepository(_DbMock.Object);
+            var res = await repository.GetByIdAsync(id);
+            Assert.IsNull(res);
+        }
+
+        [Test]
+        [TestCase(1)]
+        public async System.Threading.Tasks.Task GetByIdExisting(int id)
```
Comment (PushMeNow) [inline DIMS-Core/DIMS-Core.Tests/DAL/TaskRepositoryTest.cs:75]: namespace

File: DIMS-Core/DIMS-Core.Tests/DAL/DirectionRepositoryTest.cs (hunk 3 above)
Comment (darkpchela) [reply DIMS-Core/DIMS-Core.Tests/DAL/DirectionRepositoryTest.cs:37]: I think, i need to use it, because type "Task" already exists as type of entity in ~.DataAccessLayer.Entities

File: DIMS-Core/DIMS-Core.Tests/DAL/DirectionRepositoryTest.cs [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+        [Test]
+        public async System.Threading.Tasks.Task CreateNewEntity()
+        {
+            InitializeDbWithFourObjects();
+            Direction direction = new Direction
+            {
+                Name = "C++",
+                Description = "none"
+            };
+            var repo = new DirectionRepository(_DbMock.Object);
```
Comment (PushMeNow) [inline DIMS-Core/DIMS-Core.Tests/DAL/DirectionRepositoryTest.cs:29]: You are using repo in each method. Move this to constructor.

File: DIMS-Core/DIMS-Core.Tests/DAL/DirectionRepositoryTest.cs [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+            _DbSetDirectionMock.Verify(m => m.Remove(It.IsAny<Direction>()), Times.Once);
+        }
+
+        [Test]
+        public void GetAllFromRepository()
+        {
+            InitializeDbWithFourObjects();
+            var repo = new DirectionRepository(_DbMock.Object);
+            var res = repo.GetAll();
+            Assert.IsTrue(res.Count() == 4, res.Count().ToString());
```
Comment (PushMeNow) [inline DIMS-Core/DIMS-Core.Tests/DAL/DirectionRepositoryTest.cs:51]: You cannot be sure GetAll method returns you 4 items. Again what will happen if quantity of items is changed?

File: DIMS-Core/DIMS-Core.Tests/DAL/DirectionRepositoryTest.cs [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+
+        [Test]
+        [TestCase(1)]
+        public async System.Threading.Tasks.Task GetByIdExisting(int id)
+        {
+            InitializeDbWithFourObjects();
+            _DbSetDirectionMock.Setup(m => m.FindAsync(It.IsAny<int>())).ReturnsAsync(() => _DbSetList.SingleOrDefault(d => d.DirectionId == id));
+            var repo = new DirectionRepository(_DbMock.Object);
+            var res = await repo.GetByIdAsync(id);
+            Assert.AreEqual(_DbSetList.ElementAt(id - 1), res);
```
Comment (PushMeNow) [inline DIMS-Core/DIMS-Core.Tests/DAL/DirectionRepositoryTest.cs:62]: ElementAt? It is bad way because you cannot be sure items of list is orderable.

File: DIMS-Core/DIMS-Core.Tests/DAL/TaskRepositoryTest.cs [hunk 16]
```diff
@@ -0,0 +1,114 @@
+﻿using DIMS_Core.DataAccessLayer.Context;
```
Comment (PushMeNow) [inline DIMS-Core/DIMS-Core.Tests/DAL/TaskRepositoryTest.cs:1]: All comments about repo, naming etc  DirectionRepositoryTest.cs are related with this class too.

File: DIMS-Core/DIMS-Core.Tests/DAL/TaskStateRepositoryTest.cs [hunk 17]
```diff
@@ -0,0 +1,101 @@
+﻿using DIMS_Core.DataAccessLayer.Context;
```
Comment (PushMeNow) [inline DIMS-Core/DIMS-Core.Tests/DAL/TaskStateRepositoryTest.cs:1]: All comments about repo, naming etc  DirectionRepositoryTest.cs are related with this class too.
